sprinkle/user-facing/pytorch_trial.md

```diff
@@ -0,0 +1,240 @@
+"""
+
+import torch
+from torch import nn
+
+import determined as det
+import determined.pytorch
+
+## Data Definition ##
+
+class OnesDataset(torch.utils.data.Dataset):
+    def __len__(self):
+        return 64
+
+    def __getitem__(self, index):
+        # return data, labels
+        return torch.Tensor([1.0]), torch.Tensor([1.0])
+
+## Model Definition ##
+
+class OneVarPytorchTrial(det.pytorch.PyTorchTrial):
+    def __init__(self, context):
+        # Watch out! This is a trial context, not a trainer context like below!
+        self.context = context
+
+        # Wrap any models with context.wrap_model()
+        self.model = context.wrap_model(nn.Linear(1, 1, False))
+
+        lr = context.hparams["learning_rate"]
+
+        # Wrap any optimizers with context.wrap_optimizer()
+        self.opt = context.wrap_optimizer(
+            torch.optim.SGD(self.model.parameters(), lr=lr)
+        )
+
+    def train_batch(self, batch, epoch_idx, batch_idx):
+        """
+        Tell us how to calculate the loss/metrics for a single batch.
+        """
+        data, labels = batch
+        preds = self.model(batch)
+        loss = torch.nn.MSELoss()(preds, labels)
+        # use context.backward() and context.step_optimizer()
+        self.context.backward(loss)
+        self.context.step_optimizer(self.opt)
+        return {"loss": loss}
+
+    def evaluate_batch(self, batch: pytorch.TorchData):
+        """
+        Tell us how to calculate validation metrics for a single batch.
+        """
+        data, labels = batch
+        preds = self.model(data)
+        loss = torch.nn.MSELoss()(preds, labels)
+        return {"loss": loss}
+
+    # you can also define build_training_data_loader() or
+    # build_validation_data_loader() as methods of your PyTorchTrial,
+    # but in this example we show how to pass the data loaders in via
+    # the trainer API.
+
+## Training script ##
+
+if __name__ == "__main__":
+
+    # Watch out! This is a trainer context, not a trial context like above!
+    # (this step will also initialize random seeds automatically)
+    context = det.pytorch.init()
+
+    # Access your hparams from the trainer context.
+    hparams = context.training.hparams
+    global_batch_size = context.training.hparams["global_batch_size"]
+
+    # Optional for distributed training: cacluate this worker's
+    # batch_size so that the effective batch_size across all workers
+    # remains constant.  If you choose not to use this, the
+    # effective batch_size will be multiplied by the number slots_per_trial.
+    batch_size = context.distributed.shard_batch_size(global_batch_size)
+
+    # det.pytorch.DataLoader is a drop-in replacement for
+    # torch.utils.data.DataLoader for non-iterable Datasets.
+    train_data = det.pytorch.DataLoader(OnesDataset(), batch_size)
+    val_data = det.pytorch.DataLoader(OnesDataset(), batch_size)
+
+    # Create your Trial class.
+    my_trial = context.build_trial(OneVarPytorchTrial)
```

Comment:
Can we allow the user to define the optimizer/model outside of the trial ?

Reply:
Um, I'd have to think about it, but I think the answer would be no.  You need some boundaries on what to save into checkpoints, and you need a very well-defined order of how objects are instantiated to make it reliable to load a checkpoint.  That's why PyTorchTrial still seems very important to the high-level pytorch sprinkle api.